### ISSUE TITLE
First-enumerated GPU only: RCCL all_gather silently half-wrong over SHM, and a TCP write to address 0 over the socket transport (gfx1030 V620, 8 GPUs behind 2 PCIe switches, no P2P)

### ISSUE THREAD

**Issue by Omnicef** [ROCm/rocm-systems#12145]
## Summary

This looks like the same symptom family as **ROCm/legacy-rocm-build#6565** (formerly ROCm/ROCm#6565, still open): all_gather silently returns wrong data while all_reduce is correct, with the corruption on one channel's half. There, the channel-0 `devRingUserRanks` H2D copy during communicator init is not yet in device memory after the stream synchronize. #6565 was reported against RCCL 2.27.7 / 2.28.3, and a second machine did not reproduce it on 2.30.4. **It still reproduces here on GPU 0 with ROCm 7.14.1 / RCCL `2.30.4-HEAD:ca887ee`**, whose `projects/rccl/src/init.cc` has no readback of `devRingUserRanks` (neither does current `develop`). #6565's workaround `HSA_DISABLE_CACHE=1` makes ours clean too (2 of 2 runs). **ROCm/rocm-systems#11693** (freshly mapped VM memory reads back as zeros because the map returns before it is visible to GPU work) looks like the likely lower-layer relative.

What is new here: on our 8x Radeon Pro V620 (gfx1030) host, it happens on **exactly one GPU: whichever enumerates first** (physical device 0, the lowest PCI bus), whatever card or slot that is (we moved cards twice to establish this).

- **RCCL 2.30.4 (TheRock 7.14.1), rccl-tests single process, SHM default:** `all_gather_perf -c 1` fails (**#wrong 65536 at 256 KB, 2,097,150 at 8 MB; "Out of bounds values : 4 FAILED"**). No PyTorch involved.
- **RCCL 2.27.7 (ROCm 7.2.1), same test:** clean in a single process; the multi-process (PyTorch, SHM) path still gives the GPU-0 rank 200/200 rounds with exactly 50 % of its elements wrong.
- **Socket transport (`NCCL_SHM_DISABLE=1`), either RCCL:** GPU 0 takes a **gfxhub page fault: a TCP write to address 0x0**. With 2.30.4, rccl-tests segfaults with `Memory Fault Error [... GPU index: 0, faulting addr: 0x0, kernel: ncclDevKernel_Generic_4(...)]`. With 2.27.7 it faults for `all_reduce_perf` too.
- **`HSA_DISABLE_CACHE=1`:** GPU 0 clean (the 200-round harness, twice).
- Every group that does not include GPU 0 is clean, with every transport.

## Reproducers

**R1 (no PyTorch, no special env; silent corruption).** rccl-tests `40b1b17` built against TheRock ROCm 7.14.1 (`RCCL version : 2.30.4-HEAD:ca887ee`):

```
ROCR_VISIBLE_DEVICES=0,1,2,3 NCCL_P2P_DISABLE=1 \
  ./all_gather_perf -b 256K -e 8M -f 32 -g 4 -d half -c 1 -n 20 -w 5
#       size         count      type   redop    root     time   algbw   busbw #wrong ...
      262144         32768      half    none      -1    103.3    2.54    1.90  65536 ...
     8388608       1048576      half    none      -1   2559.1    3.28    2.46  2.09714e+06 ...
# Out of bounds values : 4 FAILED
```
**Control:** the same 2.30.4 binary with GPU 0 hidden (`ROCR_VISIBLE_DEVICES=1,2,3`, `-g 3`, the other three GPUs on the same switch) passes: `#wrong 0` at both sizes, `Out of bounds values : 0 OK`. The same test built against ROCm 7.2.1 (RCCL 2.27.7-HEAD:96a25b5) passes on the GPU-0 set and on the other switch's four GPUs.

**R1b (hard fault).** The same command with `NCCL_SHM_DISABLE=1`:
```
:0:rocdevice.cpp :3678: Memory Fault Error [..., GPU index: 0, faulting addr: 0x0, kernel: ncclDevKernel_Generic_4(ncclDevKernelArgsStorage<4096ul>)]
amdgpu 0000:07:00.0: amdgpu: [gfxhub] page fault (src_id:0 ring:40 vmid:8 pasid:..., for process all_gather_perf ...)
amdgpu 0000:07:00.0: amdgpu:   in page starting at address 0x0000000000000000 from client 0x1b (UTCL2)
amdgpu 0000:07:00.0: amdgpu:          Faulty UTCL2 client ID: TCP (0x8)
amdgpu 0000:07:00.0: amdgpu:          PERMISSION_FAULTS: 0x5
amdgpu 0000:07:00.0: amdgpu:          RW: 0x1
```
With RCCL 2.27.7 the same `NCCL_SHM_DISABLE=1` run faults the same way, for `all_reduce_perf` as well.

**R2 (multi-process, SHM default; silent corruption).** `torchrun --nproc-per-node 4`, TheRock ROCm 7.14.1 image (torch 2.12, NCCL API 2.30.4), `NCCL_P2P_DISABLE=1`:

```python
src = torch.full((1,), float(rank + 1), device=f"cuda:{rank}")
dst = torch.zeros(world, device=f"cuda:{rank}")
dist.all_gather_into_tensor(dst, src)          # GPU-0 rank gets [2.0, 0.0, 0.0, 0.0]; others [1,2,3,4]
```

A 200-round harness (48 queued matmul+all_gather pairs, M=8 and M=256 rows × 4096 fp16) fails 200/200 on the GPU-0 rank with 50 % of its elements wrong, at world size 4 and 8.

## Pattern

In the first collective of a probe (M=8 rows × 4096 float32; each value encodes its source rank and index; the output is pre-filled with a sentinel), world 4 and world 8 alike: on the GPU-0 rank, **exactly the first half of the affected slices is wrong and every second half is right**. Its own slice's first half holds **rank 1's first half at the same offsets**, and the other ranks' first halves are **never written** (sentinel intact). At world 8, the slices from ranks 1-3 (same switch) were correct and ranks 4-7's first halves were unwritten. That fits one of RCCL's two channels delivering nothing, or the wrong thing, into GPU 0's buffer.

## It follows the first-enumerated GPU, not a card, slot or rank

200-round all_gather harness; "bad" = 200/200 rounds, 50 % of that rank's elements.

| # | layout (slot -> bus, card serial) | ranks | result |
|---|---|---|---|
| 1 | SLOT1 -> bus 06 (card ...519) = GPU 0 | 06 at rank 0, PLX1 set | bus 06 bad |
| 2 | same | 06 at rank 1 | bus 06 bad, rank 0 clean |
| 3 | same | PLX2 set (no GPU 0) | clean |
| 4 | same | 06 + 2 PLX2 cards (cross-switch) | bus 06 bad |
| 5 | same | 06 at rank 2, cross-switch | bus 06 bad |
| 6 | same | all 8, 06 at rank 0 | bus 06 bad, ranks 1-7 clean |
| 7 | cards swapped: SLOT1 -> bus 06 (card ...428) = GPU 0 | 06 at rank 0 | bus 06 bad (the card that was clean in #3) |
| 8 | card ...519 now in SLOT9 -> bus 19 | 19 at rank 1 and at rank 0 | clean (the card that was bad in #1-6) |
| 9 | card ...428 moved to SLOT5 -> bus 07, SLOT1 **empty** -> bus 07 = GPU 0 | 07 at rank 0 and at rank 1 | bus 07 bad |
| 10 | same layout, rccl-tests on RCCL 2.30.4, single process | 07/0a/0d/10 (GPU 0 in set) vs 0a/0d/10 (GPU 0 hidden) | **FAILED** (#wrong 50 %) vs clean (#wrong 0) |

The failing rank is always the one on physical GPU 0, whatever card or slot that is, whichever rank it holds, at world 4 or 8, and whether or not the group crosses switches. Hiding GPU 0 with `ROCR_VISIBLE_DEVICES` makes every group clean.

## Controls

- **`HSA_DISABLE_CACHE=1` (the #6565 workaround):** GPU 0 clean, 200-round harness, 2 of 2 runs, both sizes, every rank.
- **SDMA:** `HSA_ENABLE_SDMA=0` does not help (GPU 0 still 200/200 bad; a clean set with SDMA off stays clean). The setting took effect: 256 MB D2H copies went from 13.7 to 10.4 GB/s.
- **Firmware:** `vbios_version` `113-D6030500-100` and `amdgpu_firmware_info` are byte-identical on all 8 cards (ME/PFP/CE 44, MEC 0x74, RLC 0x60, SDMA 52/0x53, SMC 58.89.0, SOS 0x00210e64).
- **Config space:** the PCIe path of the GPU-0 slot and of a clean slot are identical once addresses are masked (DevCtl/DevCtl2, MaxPayload 256 / MaxReadReq 128, LnkCtl/LnkCtl2, bridge control, ACS); only window base addresses differ. Links are 8 GT/s x16 at the slot and 16 GT/s x16 on-package.
- **P2P:** not available on this host (`hipDeviceCanAccessPeer` = 0, `hipDeviceEnablePeerAccess` = 101); `NCCL_P2P_DISABLE=1` in every run; RCCL reports `via SHM/direct/direct`.

## Host

- Supermicro SYS-4028GR-TRT2, 2x Xeon E5-2690 v4 (**Broadwell-EP**, host bridge 8086:6f00), 251 GB RAM.
- 8x Radeon Pro V620 (Navi 21, **gfx1030**, 32 GB, large BAR 32 GB) behind two **PLX 9797** switches on root ports 00:02.0 and 00:03.0 (4 GPUs each).
- Kernel 6.8.0-139-generic (Ubuntu 24.04), amdgpu built from the same source with only a power-limit-floor patch; `amdgpu.ras_enable=0 amdgpu.ppfeaturemask=0xfff73fff pci=realloc=off`. ACS redirect on (firmware default 0x001d).
- **IOMMU translated** (Intel IOMMU on by default, every group `DMA-FQ`, not passthrough).
- User space: TheRock ROCm 7.14.1 (`RCCL version : 2.30.4-HEAD:ca887ee`, torch 2.12.0) and ROCm 7.2.1 (`RCCL version : 2.27.7-HEAD:96a25b5`); both reproduce (2.27.7 not in the single-process SHM case).

## What we can run next

The host is available for more tests: `NCCL_DEBUG=INFO` / `NCCL_DEBUG_SUBSYS=ALL` on the GPU-0 rank vs a clean rank, `HSA_*` / `HIP_*` debug knobs, a KFD topology dump, or a build with extra logging. We now keep GPU 0 out of every collective group, so we can test without affecting our service.

**Comment by Omnicef** [comment]
**Does not reproduce on Ubuntu HWE 7.0.0-34 (in-tree amdgpu 3.64.0), IOMMU translated.** Same host, same card layout, same containers and binaries. The only change was a one-shot boot into `7.0.0-34-generic` instead of `6.8.0-139-generic`. amdgpu is built from Ubuntu's 7.0 source with only our power-limit-floor patch; kernel parameters and IOMMU mode are unchanged.

| test (GPU 0 = physical device 0) | 6.8.0-139, earlier the same day | 7.0.0-34 |
|---|---|---|
| rccl-tests `all_gather_perf -g 4 -c 1`, RCCL 2.30.4-HEAD:ca887ee, GPU 0 in the set | `#wrong` 50 %, Out of bounds 4 FAILED | **`#wrong 0`, Out of bounds 0 OK** |
| same, GPU 0 hidden (control) | clean | clean |
| 200-round torch all_gather harness, GPU 0 at rank 0 | 200/200 bad on the GPU-0 rank | **0/200 on every rank** |
| gfxhub page faults (boot, arms, 3 vLLM TP4 starts) | faults with `NCCL_SHM_DISABLE=1` | **0** |

So the fault is below RCCL, somewhere between Ubuntu 6.8's amdgpu/KFD and 7.0's. One boot so far; we will keep serving on 7.0 and report back if it returns. If you can point at the amdgpu/KFD change that fixed it, we can confirm with a bisect.
